### DISCUSSION TITLE
Index on a nested dictionary, where to start?

### DISCUSSION THREAD

**Discussion by famschopman** [ravendb/ravendb#13718] [Q&A] [answered]
I am looking for document/examples/pointers on creating an index where the information required is nested in dictionaries. Each schema is stored as a document. Database, Table and Column are not individually stored.

The end goal is to have an index containing the schemaId, databaseId, tableId, columnId and columnName to enable a search for a columnName and then quickly retrieve all relevant dependencies without having to open each Schema document individually.

```
public class Schema{
public Dictionary<string, Database> Databases { get; set; }
}
public class Database{
public Dictionary<string, Table> Tables{ get; set; }
}
public class Table{
public Dictionary<string, Column> Columns{ get; set; }
}
```

Hope someone can provide me show me where to find the fishing rod, so I can catch the fish myself.

I did try with the following index, but was not surprised it gave an error. It was more an attempt of trial and error.

```
public class AllFieldsInSchemas : AbstractIndexCreationTask<Schema>
    {
        public AllFieldsInSchemas()
        {
            Map = schemas => from schema in schemas
                             select new
                             {
                                 SchemaId = schema.Id,
                                 DatabaseId = schema.Databases.Select(x => x.Value.Id).ToArray(),
                                 TableId = schema.Databases.SelectMany(d => d.Value.Tables.Select(t => t.Value.Id)).ToArray(),
                                 ColumnId = schema.Databases.SelectMany(d => d.Value.Tables.SelectMany(t => t.Value.Columns.Select(c => c.Value.Id)).ToArray()),
                                 ColumnName = schema.Databases.SelectMany(d => d.Value.Tables.SelectMany(t => t.Value.Columns.Select(c => c.Value.Name)).ToArray())
                             };
                             
        }
    }


```

**Comment by haludi** [comment, marked as answer]
Hi,

You can use the fanout index -
https://ravendb.net/docs/article-page/4.2/csharp/indexes/fanout-indexes
You can also store some of the data in the index and retrieve it by
projection.
https://ravendb.net/docs/article-page/4.2/csharp/indexes/storing-data-in-index
https://ravendb.net/docs/article-page/5.0/Java/indexes/querying/projections
Be aware, storing data in the index increases the size of the index
(relevant for the memory and disk space).

Best regards,
Igal
